Sample/EventsVersioning/EventsVersioning.Tests/ExplicitSerialization/ShoppingCart.cs

```diff
@@ -0,0 +1,413 @@
+
+        return new Price(value);
+    }
+}
+
+public abstract record ShoppingCartEvent
+{
+    public record ShoppingCartOpened(
+        ShoppingCartId ShoppingCartId,
+        ClientId ClientId
+    ): ShoppingCartEvent;
+
+    public record ProductItemAddedToShoppingCart(
+        ShoppingCartId ShoppingCartId,
+        PricedProductItem ProductItem
+    ): ShoppingCartEvent;
+
+    public record ProductItemRemovedFromShoppingCart(
+        ShoppingCartId ShoppingCartId,
+        PricedProductItem ProductItem
+    ): ShoppingCartEvent;
+
+    public record ShoppingCartConfirmed(
+        ShoppingCartId ShoppingCartId,
+        LocalDateTime ConfirmedAt
+    ): ShoppingCartEvent;
+
+    public record ShoppingCartCanceled(
+        ShoppingCartId ShoppingCartId,
+        LocalDateTime CanceledAt
+    ): ShoppingCartEvent;
+}
+
+public enum ShoppingCartStatus
+{
+    Pending = 1,
+    Confirmed = 2,
+    Canceled = 4,
+
+    Closed = Confirmed | Canceled
+}
+
+public record PricedProductItem(
+    ProductId ProductId,
+    Quantity Quantity,
+    Price UnitPrice
+);
+
+public record ShoppingCart(
+    ShoppingCartId Id,
+    ClientId ClientId,
+    ShoppingCartStatus Status,
+    Dictionary<ProductId, Quantity> ProductItems
+)
+{
+    public static ShoppingCart When(ShoppingCart entity, object @event)
+    {
+        return @event switch
+        {
+            ShoppingCartOpened (var cartId, var clientId) =>
+                entity with
+                {
+                    Id = cartId,
+                    ClientId = clientId,
+                    Status = ShoppingCartStatus.Pending,
+                    ProductItems = new Dictionary<ProductId, Quantity>()
+                },
+
+            ProductItemAddedToShoppingCart (_, var productItem) =>
+                entity with { ProductItems = entity.ProductItems.Add(productItem) },
+
+            ProductItemRemovedFromShoppingCart (_, var productItem) =>
+                entity with { ProductItems = entity.ProductItems.Remove(productItem) },
+
+            ShoppingCartConfirmed (_, var confirmedAt) =>
+                entity with { Status = ShoppingCartStatus.Confirmed },
+
+            ShoppingCartCanceled (_, var canceledAt) =>
+                entity with { Status = ShoppingCartStatus.Canceled },
+            _ => entity
+        };
+    }
+
+    public static ShoppingCart Default() =>
+        new(ShoppingCartId.Unknown, ClientId.Unknown, default, new Dictionary<ProductId, Quantity>());
+}
+
+public static class ProductItemsExtensions
+{
+    public static Dictionary<ProductId, Quantity> Add(this Dictionary<ProductId, Quantity> productItems,
+        PricedProductItem productItem) =>
+        productItems
+            .Union(new[] { new KeyValuePair<ProductId, Quantity>(productItem.ProductId, productItem.Quantity) })
+            .GroupBy(ks => ks.Key)
+            .ToDictionary(ks => ks.Key, ps => Quantity.Parse((uint)ps.Sum(x => x.Value.Value)));
```

Comment:
would be much clearer to have an AddOrUpdate helper a la ConcurrentDictionary and then have 2 simple bits of code that either add to current value or insert the value - thats why its a dictionary isnt it?

Reply:
That makes sense, although then I'd need to explain it in the article 😅

Reply:
I think they'd survive - you want to tease them into looking at the code on the github!